nucleus/v1/types.go

```diff
@@ -0,0 +1,187 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NucleusHubList is a collection of nucleus hub.
+type NucleusHubList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is a list of nucleus hub.
+	Items []NucleusHub `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// NucleusHubComponents represents the components to be installed for nucleus hub.
+type NucleusHubComponents struct {
+	// Registration represents the desired configuration of registration controller installed on hub.
+	// +required
+	Registration RegistrationComponent `json:"registration" protobuf:"bytes,1,opt,name=registration"`
+}
+
+// RegistrationComponent represents the desired configuration of registration controller installed on hub.
+type RegistrationComponent struct {
+	CommonComponentConfig `json:",inline" protobuf:"bytes,1,opt,name=commonComponentConfig"`
+}
+
+// CommonComponentConfig represents the common configuration of the component.
+type CommonComponentConfig struct {
+	// Replicas is the number of desired replicas
+	Replicas int32 `json:"replica,omitempty" protobuf:"varint,1,opt,name=replicas"`
+	// Image is the image name of the component. If not set, default imagee is used.
+	// +optional
+	Image string `json:"image,omitempty" protobuf:"bytes,2,opt,name=image"`
+
+	// ImagePullPolicy is the policy to pull image.
+	// +optional
+	ImagePullPolicy corev1.PullPolicy `json:"pullPolicy,omitempty" protobuf:"bytes,3,opt,name=imagePullPolicy"`
+
+	// ResourceRequirements represents the resource requirement of the component.
+	// +optional
+	ResourceRequirements corev1.ResourceRequirements `json:"resourceRequirements,omitempty" protobuf:"bytes,4,opt,name=resourceRequirements"`
+}
+
+// WorkComponent represeents the desired configuration of work component.
+type WorkComponent struct {
+	CommonComponentConfig `json:",inline" protobuf:"bytes,3,opt,name=commonComponentConfig"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+
+// NucleusAgent represents a deployment of nucleus agent on spoke cluster.
+type NucleusAgent struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,3,opt,name=metadata"`
+
+	// Spec represents the desired deployment configuratioin of nucleus agent.
+	Spec NucleusAgentSpec `json:"spec,omitempty" protobuf:"bytes,1,opt,name=spec"`
+
+	// Status represents the current status of nucleus agent.
+	Status NucleusAgentStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+
+// NucleusAgentSpec represents the desired deployment configuratioin of nucleus agent.
+type NucleusAgentSpec struct {
```

Comment:
to allow connections to multiple hubs, you must have a namespace.  Choose some name restrictions it (prefix) and a default value, but I suspect an optional namespace will be important.

Reply:
I think that means the the api (also nucleushub api) should be cluster scoped. And nucleushub api should be singleton?

Reply:
I think that means that we should:

- pick a default namespace
- restrict names to be the default or an optional suffix on the default
- allow specification of the namespace in the API

I'm not sure my @deads2k model could predict whether this would yield a namespaced resource or not. I would _guess_ cluster scoped.